### PR TITLE
cache reconciliation on leader failover

### PR DESCRIPTION
**Pull request by onursatici**

### CODE REVIEW

File: internal/extender/failover.go [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+		return nil, nil, err
+	}
+
+	nodes, nodesFound := r.orderedNodes[instanceGroup]
+	availableResources, resourcesFound := r.availableResources[instanceGroup]
+	if !nodesFound || !resourcesFound {
+		return nil, nil, werror.Error("instance group not found", werror.SafeParam("instanceGroup", instanceGroup))
+	}
+
+	executorCountToReserve := applicationResources.executorCount - len(executors)
```
Comment (rkaram) [inline internal/extender/failover.go]: Don't you want to create reservations for stale executors instead?

File: internal/extender/failover.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+		// the driver is stale, a new resource reservation object needs to be created
+		instanceGroup := sp.driver.Spec.NodeSelector[instanceGroupNodeSelector]
+		newRR, reservedResources, err := r.constructResourceReservation(ctx, sp.driver, sp.executors, instanceGroup)
+		if err != nil {
+			svc1log.FromContext(ctx).Error("failed to construct resource reservation", svc1log.Stacktrace(err))
+			return
+		}
+		err = r.resourceReservations.Create(newRR)
+		if err != nil {
+			logRR(ctx, "resource reservation already exists, force updating", sp.driver.Namespace, sp.driver.Labels[SparkAppIDLabel])
```
Comment (rkaram) [inline internal/extender/failover.go]: Just curious in which scenario you're expecting this to happen...

File: internal/extender/failover.go [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+}
+
+func (r *reconciliator) deleteDemandIfExists(namespace, name string) {
+	_, ok := r.demands.Get(namespace, name)
+	if ok {
+		r.demands.Delete(namespace, name)
+	}
+}
+
+func sparkPodsByAppID(
```
Comment (rkaram) [inline internal/extender/failover.go]: nit, but I think this should either have a name or documentation that clarifies it drops already reserved pods.

File: internal/logging.go [hunk 4]
```diff
@@ -35,6 +36,14 @@ func DemandSafeParams(demandName string, demandNamespace string) map[string]inte
 	}
 }
 
+// ResourceReservationSafeParamsFromObj gets the safe params for a resource reservation object
+func ResourceReservationSafeParamsFromObj(rr *v1beta1.ResourceReservation) map[string]interface{} {
```
Comment (rkaram) [inline internal/logging.go]: This does not seem to be used

File: internal/cache/safedemands.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+func (sdc *SafeDemandCache) IsLoaded() bool {
+	return sdc.demandCRDInitialized.Load()
+}
+
+// Create enqueues a creation request and puts the object into the store
+func (sdc *SafeDemandCache) Create(rr *demandapi.Demand) error {
+	if !sdc.demandCRDInitialized.Load() {
+		return nil
```
Comment (laflechejonathan) [inline internal/cache/safedemands.go]: should return an error

File: internal/extender/demand.go [hunk 6]
```diff
@@ -60,7 +56,7 @@ func (s *SparkSchedulerExtender) updatePodStatus(ctx context.Context, pod *v1.Po
 }
 
 func (s *SparkSchedulerExtender) createDemandForExecutor(ctx context.Context, executorPod *v1.Pod, executorResources *resources.Resources) {
-	if !s.demandCRDInitialized.Load() {
+	if !s.demands.IsLoaded() {
```
Comment (laflechejonathan) [inline internal/extender/demand.go]: s/loaded/exists/ or registered

File: internal/extender/failover.go [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+	"github.com/palantir/witchcraft-go-logging/wlog/svclog/svc1log"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	corelisters "k8s.io/client-go/listers/core/v1"
+)
+
+// SyncResourceReservationsAndDemands gets all resource reservations and pods,
+// finds all pods that do not have a claimed spot in their resource reservation object,
+// groups them by spark application id, and patched the resource reservation objects so
```
Comment (laflechejonathan) [inline internal/extender/failover.go]: *patches

File: internal/extender/failover.go (hunk 3 above)
Comment (laflechejonathan) [reply internal/extender/failover.go]: something like `getUnreservedSparkPodsBySparkID`

File: internal/extender/failover.go [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+	// if the driver is nil it already has an associated reservation, get the resource
+	// reservation object and update it so it has reservations for each stale executor
+	if sp.driver == nil && len(sp.executors) > 0 {
+		exec := sp.executors[0]
+		rr, ok := r.resourceReservations.Get(exec.Namespace, exec.Labels[SparkAppIDLabel])
+		if !ok {
+			logRR(ctx, "resource reservation deleted, ignoring", exec.Namespace, exec.Labels[SparkAppIDLabel])
+			return
+		}
+		err := r.patchResourceReservation(sp.executors, rr.DeepCopy())
```
Comment (laflechejonathan) [inline internal/extender/failover.go]: not clear to me why you need a Deep Copy

File: internal/extender/failover.go [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+	availableResources   map[string]resources.NodeGroupResources
+	orderedNodes         map[string][]*v1.Node
+}
+
+func (r *reconciliator) syncResourceReservation(ctx context.Context, sp *sparkPods) {
+	// if the driver is nil it already has an associated reservation, get the resource
+	// reservation object and update it so it has reservations for each stale executor
+	if sp.driver == nil && len(sp.executors) > 0 {
+		exec := sp.executors[0]
+		rr, ok := r.resourceReservations.Get(exec.Namespace, exec.Labels[SparkAppIDLabel])
```
Comment (laflechejonathan) [inline internal/extender/failover.go]: I think your code will be cleaner if `sparkPods` had a SparkAppId field.

File: internal/extender/failover.go [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+	}
+	// recompute overhead to account for newly created resource reservations
+	overheadComputer.compute(ctx)
+	return nil
+}
+
+// sparkPods is a collection of stale state, it is comprised of
+// pods from a spark application that do not have a claimed resource reservation,
+// and the last known state of the resource reservation object
+type sparkPods struct {
```
Comment (laflechejonathan) [inline internal/extender/failover.go]: rename to something like `inconsistentSparkApplication`

Comment (laflechejonathan) [reply internal/extender/failover.go]: or actually, do this:
```golang
type sparkApplication struct {
    sparkAppID string
    inconsistentDriver *v1.Pod
    inconsistentExecutors []*v1.Pod
}```

File: internal/extender/failover.go [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+				logRR(ctx, "resource reservation deleted, ignoring", sp.driver.Namespace, sp.driver.Labels[SparkAppIDLabel])
+				return
+			}
+		}
+		r.availableResources[instanceGroup].Sub(reservedResources)
+	}
+}
+
+func (r *reconciliator) syncDemand(ctx context.Context, sp *sparkPods) {
+	// TODO check if demand crd exists
```
Comment (laflechejonathan) [inline internal/extender/failover.go]: rm

File: internal/extender/failover.go (hunk 2 above)
Comment (onursatici) [reply internal/extender/failover.go]: Not expecting this to happen, but want to guard against two extenders starting up. What I will do is to make sure we run only on the leader on startup

File: internal/extender/failover.go (hunk 1 above)
Comment (onursatici) [reply internal/extender/failover.go]: I think the wording is off here, these executors do not exist yet so we need to reserve spaces for them, as opposed to `executors` here that exist but do not have the corresponding resource reservation object

File: internal/extender/failover.go (hunk 8 above)
Comment (onursatici) [reply internal/extender/failover.go]: you can't change objects you got from the informer / cache

**Pull request merged**